logziosender.go

```diff
@@ -262,24 +315,33 @@ func (l *LogzioSender) Drain() {
 	defer l.draining.Toggle()
 
 	l.buf.Reset()
-	bufSize := l.dequeueUpToMaxBatchSize()
-	if bufSize > 0 {
-		backOff := sendSleepingBackoff
-		toBackOff := false
-		for attempt := 0; attempt < sendRetries; attempt++ {
-			if toBackOff {
-				l.debugLog("logziosender.go: failed to send logs, trying again in %v\n", backOff)
-				time.Sleep(backOff)
-				backOff *= 2
-			}
-			statusCode := l.tryToSendLogs()
-			if l.shouldRetry(attempt, statusCode) {
-				toBackOff = true
-			} else {
-				break
+	var reDrain bool = true
+	for l.queue.Length() > 0 && reDrain {
+		bufSize := l.dequeueUpToMaxBatchSize()
+		if bufSize > 0 {
+			backOff := sendSleepingBackoff
+			toBackOff := false
+			for attempt := 0; attempt < sendRetries; attempt++ {
+				if toBackOff {
+					l.debugLog("logziosender.go: failed to send logs, trying again in %v\n", backOff)
+					time.Sleep(backOff)
+					backOff *= 2
+				}
+				statusCode := l.tryToSendLogs()
+				if l.shouldRetry(statusCode) {
+					toBackOff = true
+					if attempt == (sendRetries - 1) {
+						l.requeue()
+						reDrain = false
+					}
+				} else {
+					reDrain = true
+					break
+				}
 			}
 		}
 	}
+
 }
 
 func (l *LogzioSender) dequeueUpToMaxBatchSize() int {
```

Comment:
few questions on this functions
1. where do you get the item value from ?
2. Its look like you are getting only 1 item - where is the batch?

Reply:
`dequeueUpToMaxBatchSize()` this method loads items to the sender buffer and only returns the buffer size